bitswap.go

```diff
@@ -325,54 +348,79 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 		return
 	}
 
-	wg := sync.WaitGroup{}
-	for _, block := range iblocks {
-
-		wg.Add(1)
-		go func(b blocks.Block) { // TODO: this probably doesnt need to be a goroutine...
-			defer wg.Done()
-
-			bs.updateReceiveCounters(b)
-			bs.sm.UpdateReceiveCounters(p, b)
-			log.Debugf("[recv] block; cid=%s, peer=%s", b.Cid(), p)
-			// skip received blocks that are not in the wantlist
-			if !bs.wm.IsWanted(b.Cid()) {
-				log.Debugf("[recv] block not in wantlist; cid=%s, peer=%s", b.Cid(), p)
-				return
-			}
-
-			if err := bs.receiveBlockFrom(b, p); err != nil {
-				log.Warningf("ReceiveMessage recvBlockFrom error: %s", err)
-			}
-			log.Event(ctx, "Bitswap.GetBlockRequest.End", b.Cid())
-		}(block)
+	bs.updateReceiveCounters(iblocks)
+	for _, b := range iblocks {
+		log.Debugf("[recv] block; cid=%s, peer=%s", b.Cid(), p)
 	}
-	wg.Wait()
-}
 
-func (bs *Bitswap) updateReceiveCounters(b blocks.Block) {
-	blkLen := len(b.RawData())
-	has, err := bs.blockstore.Has(b.Cid())
+	// Process blocks
+	err := bs.receiveBlocksFrom(p, iblocks)
 	if err != nil {
-		log.Infof("blockstore.Has error: %s", err)
+		log.Warningf("ReceiveMessage recvBlockFrom error: %s", err)
 		return
 	}
 
-	bs.allMetric.Observe(float64(blkLen))
-	if has {
-		bs.dupMetric.Observe(float64(blkLen))
+	for _, b := range iblocks {
+		if bs.wm.IsWanted(b.Cid()) {
+			log.Event(ctx, "Bitswap.GetBlockRequest.End", b.Cid())
```

Comment:
Can we push this down into `receiveBlocksFrom`? `IsWanted` is non-trivial at the moment.

Reply:
That's a little tricky because it needs a reference to the context, and `Bitswap.HasBlock()` doesn't take a context parameter. So I would need to supply a nil context. That's doable, just want to make sure you think it's worth it?

Reply:
Probably not. Ideally, we'd just optimize `IsWanted` to somehow bypass the wantmanager loop.